#19587 add yum skip-broken

vmindru opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request


##### COMPONENT NAME
modules/packaging/os/yum.py 

##### ANSIBLE VERSION
targeting 2.3  i guess


##### SUMMARY
adding skip_broken option, --skip-broken as per yum man is a general option meaning it can be added all other the place, therefore, adding it into the yum_basecmd 

bonus: did some pep8 refactor 

<!---
this fixes https://github.com/ansible/ansible/issues/19587
-->

test-module pastebin http://pastebin.com/mgCmcVdq

Commit:
#19587: refactor to PEP8

Commit:
Issue: #19587: add skip_broken option to yum module
PNTSYSOPS-1901 - internal reference

Signed-off-by: Veaceslav Mindru <vmindru@redhat.com> <mindruv@gmail.com>

Commit:
Issue: #19587: fix line break, attempt to make readble

        things at line 646 look ugly ..
        trying to make them look a bit more human readble
        though the entire approach should be rewriten

PNTSYSOPS-1901 - internal reference

Signed-off-by: Veaceslav Mindru <vmindru@redhat.com> <mindruv@gmail.com>

vmindru commented:
i hope the thingy will not complain about this 3 vars , i was thining to cover those in a separte MR 

```
[vmindru@vmutil ansible-devel]$ flake8 lib/ansible/modules/packaging/os/yum.py 
lib/ansible/modules/packaging/os/yum.py:829:13: F841 local variable 'is_group' is assigned to but never used
lib/ansible/modules/packaging/os/yum.py:1014:17: F841 local variable 'msg' is assigned to but never used
lib/ansible/modules/packaging/os/yum.py:1128:29: F841 local variable 'a' is assigned to but never used
[vmindru@vmutil ansible-devel]$ 
````

Commit:
#19587: fixing typo, skip-broken has to be passed as list

vmindru commented:
imho this is ready for review, there are this 2 issues not sure if i should remove them within this MR or as a separate one.

```[vmindru@vmutil ansible-devel]$ flake8 lib/ansible/modules/packaging/os/yum.py 
lib/ansible/modules/packaging/os/yum.py:829:13: F841 local variable 'is_group' is assigned to but never used
lib/ansible/modules/packaging/os/yum.py:1128:29: F841 local variable 'a' is assigned to but never used
[vmindru@vmutil ansible-devel]$ 
```

e.g.  IMHO this entire conditional makes no use, it's either redundant or became redundant during some change. 
```
1126                         if not i in current_repos:                                                                                                              
1127                             rid = my.repos.getRepo(i)                                                                                                           
1128                             a = rid.repoXML.repoid  ##  if no one complains remove this on next MR    
```

same goes here this peace can be rewrote
```
 825     for pkg in items:                                                                                                                                           
 826         is_group = False                                                                                                                                        
 827         # group remove - this is doom on a stick                                                                                                                
 828         if pkg.startswith('@'):                                                                                                                                 
 829             is_group = True                                                                                                                                                                                    
 830         else:                                                                                                                                                   
 831             if not is_installed(module, repoq, pkg, conf_file, en_repos=en_repos, dis_repos=dis_repos, installroot=installroot):                                
 832                 res['results'].append('%s is not installed' % pkg)                                                                                              
 833                 continue    
```
might be end-up like this
```
 825     for pkg in items and is_installed(module, repoq, pkg, conf_file, en_repos=en_repos, dis_repos=dis_repos, installroot=installroot):                          
 826         res['results'].append('%s is not installed' % pkg)                                                                                                                                                     
 827         continue    
```

vmindru commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -988,10 +1011,13 @@ def latest(module, items, repoq, yum_basecmd, conf_file, en_repos, dis_repos, in
             if w.startswith('@'):
                 to_update.append((w, None))
-                msg = '%s will be updated' % w
+                res['msg'] += '%s will be updated' % w
```
not even sure why that previous msg = '%s will be updated' % w was there, just tried to 'guess'

abadger replied:
Judging from the other things in this conditional, it probably should be appended to to_update?

vmindru replied:
in this case I think it should be removed at all cause to_updated get's updated on previous line

vmindru commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -1095,7 +1125,7 @@ def ensure(module, state, pkgs, conf_file, enablerepo, disablerepo,
                         if not i in current_repos:
                             rid = my.repos.getRepo(i)
-                            a = rid.repoXML.repoid
+                            a = rid.repoXML.repoid  ##  if no one complains remove this on next MR 
```
this is one of the var complaints, the var is assigned but not really used.

abadger replied:
Check the yum code for sideeffects (repoid might be a property and referencing it might cause some action to happen).  If this is a simply attribute rather than something producing side-effects, feel free to get rid of it.

vmindru replied:
```
  3 import yum                                                                      
  4 from json import dumps                                                          
  5                                                                                 
  6 instance = yum.YumBase()                                                        
  7 print "###################################"                                     
  8 for repo in instance.repos.listEnabled():                                       
  9   print  type(repo.repoXML.repoid)
```
says 
```
###################################
Loaded plugins: product-id
Repo rhel-7-server-satellite-tools-6.2-rpms forced skip_if_unavailable=True due to: /etc/pki/entitlement/7092233073807600698-key.pem
Repo rhel-server-rhscl-7-eus-source-rpms forced skip_if_unavailable=True due to: /etc/pki/entitlement/7092233073807600698-key.pem
Repo rhel-7-server-extras-rpms forced skip_if_unavailable=True due to: /etc/pki/entitlement/7092233073807600698-key.pem
Repo rhel-7-server-optional-rpms forced skip_if_unavailable=True due to: /etc/pki/entitlement/7092233073807600698-key.pem
Repo rhel-7-server-rpms forced skip_if_unavailable=True due to: /etc/pki/entitlement/7092233073807600698-key.pem
Repo rhel-7-fast-datapath-rpms forced skip_if_unavailable=True due to: /etc/pki/entitlement/7092233073807600698-key.pem
Repo rhel-7-server-tus-rpms forced skip_if_unavailable=True due to: /etc/pki/entitlement/7092233073807600698-key.pem
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
<type 'str'>
[vmindru@vmutil python]$ 
```
so i guess it's a string. 



```                                                                                                                                                                              
~

vmindru replied:
will remove that, it's basically repo name string.

vmindru reviewed (commented):
just couple of comments

abadger commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -631,7 +643,18 @@ def list_stuff(module, repoquerybin, conf_file, stuff, installroot='/'):
+                                                             stuff,
+                                                             conf_file,
+                                                             qf=qf,
+                                                             installroot=installroot)) if p.strip() ]
```
This is horrible formatting ;-)  If you want to break this up, make logical divisions more like this:

``` python
return [ pkg_to_dict(p) for p in sorted(
                  is_installed(module, repoq, stuff, conf_file, qf=is_installed_qf, installroot=installroot) +
                  is_available(module, repoq, stuff, conf_file, qf=qf, installroot=installroot))
            if p.strip() ]
```

vmindru replied:
ack, will do was not sure how do format this 100 organizations 100 formatting styles :-D

abadger commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -1117,23 +1147,11 @@ def ensure(module, state, pkgs, conf_file, enablerepo, disablerepo,
-    #   list=available
-    #   list=repos
-    #   list=pkgspec
-
```
Leave this, at least for now.  there's no other listing of the two types of commands (and the arguments that list takes aren't even i nthe documentatoin :-(

abadger commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -1058,6 +1084,10 @@ def ensure(module, state, pkgs, conf_file, enablerepo, disablerepo,
     en_repos = []
+
+    if skip_broken:
+        yum_basecmd.extend(['--skip-broken'])
```
Is skip-broken allowed for all of the commands that we run?  (I've used it with yum update but never with yum remove or yum install).

vmindru replied:
Yep https://linux.die.net/man/8/yum --skip-broken is in the general section meaning it can be added to any other action.

vmindru replied:
we use it quite often during Maintenance windows and mass server updates, we don't want to waste time to solve problems on failed machines, we just --skip-update and later follow-up what are the deps issues.

Commit:
PR: #21475: peer review changes
            relates to: revert comment deletion https://github.com/ansible/ansible/pull/21475#discussion_r101574146
            relates to: remove irelevant var https://github.com/ansible/ansible/pull/21475#discussion_r101573388
            relates to: reformat https://github.com/ansible/ansible/pull/21475#discussion_r101572840

Signed-off-by: Veaceslav Mindru <vmindru@redhat.com> <mindruv@gmail.com>

Commit:
PR: #21475: peer review changes
            relates to: https://github.com/ansible/ansible/pull/21475#discussion_r101573736

Signed-off-by: Veaceslav Mindru <vmindru@redhat.com> <mindruv@gmail.com>

vmindru commented:
can we ignore the CI test failure 

```
2017-02-16 17:58:13 ERROR: PEP 8: lib/ansible/modules/packaging/os/yum.py: Passes current rule set. Remove from legacy list (test/sanity/pep8/legacy-files.txt).
2017-02-16 17:58:13 ERROR: PEP 8: There are 1 issues which need to be resolved.
``` 

I presume it's caused by, and i really don't want to change that part of code 

```
[vmindru@vmutil ansible-devel]$ flake8 lib/ansible/modules/packaging/os/yum.py 
lib/ansible/modules/packaging/os/yum.py:819:13: F841 local variable 'is_group' is assigned to but never used
[vmindru@vmutil ansible-devel]$ 
```

Commit:
Issue: #21475: fixing typo in the Man section

Signed-off-by: Veaceslav Mindru <vmindru@redhat.com> <mindruv@gmail.com>

mattclay commented:
@vmindru Since you've fixed the PEP 8 issues that currently exist in the file, you need to remove the file from `test/sanity/pep8/legacy-files.txt` as part of your PR.

abadger commented on lib/ansible/modules/packaging/os/yum.py:
```diff
@@ -1084,25 +1103,18 @@ def ensure(module, state, pkgs, conf_file, enablerepo, disablerepo,
-                    for i in new_repos:
-                        if not i in current_repos:
-                            rid = my.repos.getRepo(i)
-                            a = rid.repoXML.repoid
```
I think accessing repoXML has a side effect.  This is in the yum source (yumRepo.py): 
``` python
    def _loadRepoXML(self, text=None):
        """retrieve/check/read in repomd.xml from the repository"""
        try:
            return self._groupLoadRepoXML(text, self._mdpolicy2mdtypes())
        except KeyboardInterrupt:
            self._revertOldRepoXML() # Undo metadata cookie?
            raise
        raise Errors.RepoError, 'Bad loadRepoXML policy (for %s): %s' % (self.ui_id, self.mdpolicy)

    def _getRepoXML(self):
        if self._repoXML:
            return self._repoXML
        self._loadRepoXML(text=self.ui_id)
        return self._repoXML


    repoXML = property(fget=lambda self: self._getRepoXML(),
                       fset=lambda self, val: setattr(self, "_repoXML", val),
                       fdel=lambda self: setattr(self, "_repoXML", None))
```

So we probably need to keep this.  Should add a comment to say why it's there as well.

abadger replied:
Note: If this is causing the pep8 test to complain, it would be a valid place to use the # nopep8 comment (along with the comment explaining why we have to do this).

vmindru replied:
this has been reverted, #21619 has been created not to forget to tackle this issue. I decided to take care of this issue in a separate PR to keep things less polluted.

abadger commented:
This is close.  In addition to removing from the pep8 skip file I only see one problem that needs to be addressed.

Commit:
PR: #21475 - Peer review: revert variable a
        Relates to: https://github.com/ansible/ansible/pull/21475#pullrequestreview-22404868

Signed-off-by: Veaceslav Mindru <vmindru@redhat.com> <mindruv@gmail.com>

vmindru commented:
w00t! :) all tests have passed LOL so nice to see green color all other the place. Now need final review and then I can tackle this module further under #21619

Commit:
Issue: 19587 - small correction

Signed-off-by: Veaceslav Mindru <vmindru@redhat.com> <mindruv@gmail.com>

abadger commented:
Squashed and merged to devel (for 2.3.0 release).  Thanks vmindru!
